### PR TITLE
Reader Tags Reduxification (5): reducer, actions, and tests

### PR DESCRIPTION
**Pull request by samouri**
### Table of Contents
1. new util for /read/tags data-layer: https://github.com/Automattic/wp-calypso/pull/11733
2. updating read/tags data-layer: https://github.com/Automattic/wp-calypso/pull/11627
3. unfollowing data-layer:  https://github.com/Automattic/wp-calypso/pull/11647
4. following data-layer: https://github.com/Automattic/wp-calypso/pull/11649
5. **reducers + actions: https://github.com/Automattic/wp-calypso/pull/11734**
6. query components + selectors: #11736
7. connecting: https://github.com/Automattic/wp-calypso/pull/11737
8. deleting old code: https://github.com/Automattic/wp-calypso/pull/11738

### Summary
This PR in the series that migrates tags from flux to redux.  Also this is an exciting one :).
This pr introduces the reducer for tags.  The proposes plan for the state tree is to look like:
```
 reader: { tags: { items: {
   [ tag.id ]: {
     id,
     url,
     displayName,
     title,
     slug,
     isFollowing,
   }
   ...
} } }
```

### CODE REVIEW

File: client/state/reader/tags/items/actions.js [hunk 1]
```diff
@@ -35,10 +35,10 @@ export const requestTags = tag => {
 	};
 };
 
-export const receiveTags = ( { payload, error } ) => ( {
+export const receiveTags = ( { payload, resetFollowingData = false } ) => ( {
```
Comment (samouri) [inline client/state/reader/tags/items/actions.js]: this probably belongs in (2)

Comment (samouri) [inline client/state/reader/tags/items/actions.js]: moving

Comment (samouri) [inline client/state/reader/tags/items/actions.js]: moved

File: client/state/reader/tags/items/reducer.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * we can do a full replace instead of merge
+ *
+ * the shape of a tag is { id, url, title, displayName, isFollowing }.
+ */
+export const items = createReducer( {}, {
+	[ READER_TAGS_RECEIVE ]: ( state, action ) => {
+		const tags = action.payload;
+		const resetFollowingData = action.meta.resetFollowingData;
+
+		if ( ! resetFollowingData ) {
```
Comment (blowery) [inline client/state/reader/tags/items/reducer.js]: so this assumes the tag already has `isFollowing` set?

File: client/state/reader/tags/items/reducer.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			return merge(
```
Comment (blowery) [inline client/state/reader/tags/items/reducer.js]: why a merge if we're replacing everything currently there?

File: client/state/reader/tags/items/reducer.js (hunk 2 above)
Comment (blowery) [inline client/state/reader/tags/items/reducer.js]: ohhhh, so this flag is really "assume the set of things currently being followed are represented by that tags attached to this action"

Comment (samouri) [inline client/state/reader/tags/items/reducer.js]: the flag = throw all previous following data out

File: client/state/reader/tags/items/reducer.js (hunk 3 above)
Comment (samouri) [inline client/state/reader/tags/items/reducer.js]: that was my first instinct too!
But lets say you browse to tag pages that you don't follow. Then those get cached into redux and I wouldn't want to throw them out.

Comment (blowery) [inline client/state/reader/tags/items/reducer.js]: i get it now

File: client/state/reader/tags/items/reducer.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	},
+	[ READER_UNFOLLOW_TAG_RECEIVE ]: ( state, action ) => {
+		const removedTag = action.payload;
+		return merge(
+			{},
+			state,
+			{ [ removedTag ]: { isFollowing: false } },
+		);
+	}
+} );
```
Comment (blowery) [inline client/state/reader/tags/items/reducer.js]: could we get a schema here too so we can serialize off the set of followed tags? 

might need a custom serialize / deserialize pair though to limit serialization to followed tags?

Comment (samouri) [inline client/state/reader/tags/items/reducer.js]: personally I prefer adding in serialization/deserialization in second round PRs.

**Pull request merged**